semtypes/src/main/java/io/ballerina/types/typeops/ListCommonOps.java

```diff
@@ -118,82 +119,154 @@ static boolean listFormulaIsEmpty(Context cx, Conjunction pos, Conjunction neg)
-            if (!repeatingMember.equals(members.initial.get(maxInitialLen))) {
-                members.initial.set(maxInitialLen, repeatingMember);
+        Collections.sort(fixedLengths);
+        // `boundaries` partitions the non-negative integers
+        // Construct `boundaries` from `fixedLengths` and `maxInitialLength`
+        // An index b is a boundary point if indices < b are different from indices >= b
+        //int[] boundaries = from int i in 1 ... maxInitialLength select i;
+        List<Integer> boundaries = new ArrayList<>();
+        for (int i = 1; i <= maxInitialLength; i++) {
+            boundaries.add(i);
+        }
+        for (int n : fixedLengths) {
+            // this also removes duplicates
+            if (boundaries.size() == 0 || n > boundaries.get(boundaries.size() - 1)) {
+                boundaries.add(n);
             }
         }
-        if (newTypeLen < intersectedLen) {
-            if (Core.isNever(newType.rest)) {
-                return null;
+        // Now construct the list of indices by taking nNeg samples from each partition.
+        List<Integer> indices = new ArrayList<>();
+        int lastBoundary = 0;
+        if (nNeg == 0) {
+            // this is needed for when this is used in listProj
+            nNeg = 1;
+        }
+        for (int b : boundaries) {
+            int segmentLength = b - lastBoundary;
+            // Cannot have more samples than are in the parition.
+            int nSamples = Integer.min(segmentLength, nNeg);
+            for (int i = b - nSamples; i < b; i++) {
+                indices.add(i);
             }
-            for (int i = newTypeLen; i < intersectedLen; i++) {
-                fixedArraySet(members, i, intersect(listMemberAt(members, rest, i), newType.rest));
+            lastBoundary = b;
+        }
+        for (int i = 0; i < nNeg; i++) {
+            // Be careful to avoid integer overflow.
+            if (lastBoundary > Long.MAX_VALUE - i) {
+                break;
             }
+            indices.add(lastBoundary + i);
+        }
+        return indices;
+    }
+
+    static FixedLengthArraySemtypePair listIntersectWith(FixedLengthArray members1, SemType rest1,
+                                                         FixedLengthArray members2, SemType rest2) {
+        if (listLengthsDisjoint(members1, rest1, members2, rest2)) {
+            return null;
+        }
+        ArrayList<SemType> initial = new ArrayList<>();
+        int max = Integer.max(members1.initial.size(), members2.initial.size());
+        for (int i = 0; i < max; i++) {
+            initial.add(intersect(listMemberAt(members1, rest1, i), listMemberAt(members2, rest2, i)));
         }
-        return FixedLengthArraySemtypePair.from(members, intersect(rest, newType.rest));
+        return FixedLengthArraySemtypePair.from(FixedLengthArray.from(initial,
+                Integer.max(members1.fixedLength, members2.fixedLength)), Core.intersect(rest1, rest2));
```

Comment:
```suggestion
        return FixedLengthArraySemtypePair.from(FixedLengthArray.from(initial, Integer.max(members1.fixedLength, 
                                                                                           members2.fixedLength)), 
                                                Core.intersect(rest1, rest2));
```

Reply:
Let's break the line from comma in the `Integer.max` call and align second parameter to first parameter start column of `Integer.max` call.